examples/Examples/PlutusType.hs

```diff
@@ -39,13 +40,50 @@ swap x = pmatch x $ \case
 tests :: HasTester => TestTree
 tests =
   testGroup
-    "PlutusType examples"
-    [ testCase "A encoded as 0" $ do
-        pcon A `equal` (0 :: Term s PInteger)
-    , testCase "B encoded as 2" $ do
-        pcon B `equal` (1 :: Term s PInteger)
-    , testCase "swap A == B" $ do
-        swap (pcon A) `equal` pcon B
-    , testCase "swap B == A" $ do
-        swap (pcon B) `equal` pcon A
+    "PlutusType tests"
+    [ testGroup
+        "PlutusType examples"
+        [ testCase "A encoded as 0" $ do
+            pcon A `equal` (0 :: Term s PInteger)
+        , testCase "B encoded as 2" $ do
+            pcon B `equal` (1 :: Term s PInteger)
+        , testCase "swap A == B" $ do
+            swap (pcon A) `equal` pcon B
+        , testCase "swap B == A" $ do
+            swap (pcon B) `equal` pcon A
+        ]
+    , testGroup
+        "PlutusType instances sanity"
+        -- TODO: Add more sanity tests here!
+        [ testCase "PBuiltinList" $ do
+            pmatchTargetEvalBList $ pconstant [1 :: Integer, 2, 3, 4]
+        ]
     ]
+
+-- CPP support isn't great in fourmolu.
+{- ORMOLU_DISABLE -}
+
+-- | Make sure the target builtin list of 'pmatch' is not evaluated multiple times.
+pmatchTargetEvalBList :: (HasTester, PLift a) => ClosedTerm (PBuiltinList a) -> Assertion
+pmatchTargetEvalBList target = pmatch (ptrace (pconstant tag) target) (\case
+  PCons x xs -> plet x $ \_ -> plet xs $ \_ -> pconstant ()
+  PNil -> pconstant ())
+#ifdef Development
+  `traces` replicate 1 tag
+#else
+  `traces` []
+#endif
+  where
+    tag = "evaluating"
+
+-- FIXME: This needs to fully "evaluate" (e.g evaluate in UPLC) the `x` within the `pmatch` handler.
+-- | Make sure the target of 'pmatch' is only evaluated once.
+_pmatchTargetEval :: HasTester => PlutusType p => ClosedTerm p -> Assertion
```

Comment:
This is what I was trying. It's missing one key piece- the argument to the `pmatch` handler needs to have its inner terms "evaluated" with `plet` (see: `pmatchTargetEvalBList` impl). Only then this test is valid and can work generically on any plutus type.

The thing is, I don't think we can do that without another typeclass.

Reply:
Actually never mind, I think I fixed it.